cpu/nrf52/radio/nrf802154/nrf802154.c

```diff
@@ -0,0 +1,403 @@
+}
+
+static int recv(netdev_t *dev, void *buf, size_t len, void *info)
+{
+    (void)dev;
+    (void)info;
+
+    int pktlen = (int)rxbuf[0] - 2;
+
+    /* check if packet data is readable */
+    if (pktlen <= 0) {
+        DEBUG("recv: no packet data available\n");
+        reset_rx();
+        return 0;
+    }
+
+    if (buf == NULL) {
+        if (len > 0) {
+            /* drop packet */
+            DEBUG("recv: dropping packet of length %i\n", pktlen);
+        }
+        else if (len == 0) {
+          /* return packet length */
+          DEBUG("recv: return packet length: %i\n", pktlen);
+          return pktlen;
+        }
+    }
+    else {
+        DEBUG("recv: reading packet of length %i\n", pktlen);
+        pktlen = (len < pktlen) ? len : pktlen;
+        memcpy(buf, &rxbuf[1], pktlen);
+    }
+
+    reset_rx();
+
+    return pktlen;
+}
+
+static void isr(netdev_t *dev)
+{
+    if (nrf802154_dev.netdev.event_callback) {
+        nrf802154_dev.netdev.event_callback(dev, NETDEV_EVENT_RX_COMPLETE);
+    }
+}
+
+static int get(netdev_t *dev, netopt_t opt, void *value, size_t max_len)
+{
+    assert(dev);
+
+    DEBUG("get: %s\n", netopt2str(opt));
+
+    switch (opt) {
+        case NETOPT_CHANNEL:
+            assert(max_len >= sizeof(uint16_t));
+            *((uint16_t *)value) = nrf802154_dev.chan;
+            return sizeof(uint16_t);
+        case NETOPT_TX_POWER:
+            assert(max_len >= sizeof(int16_t));
+            *((int16_t *)value) = get_txpower();
+            return sizeof(int16_t);
+
+        default:
+            return netdev_ieee802154_get((netdev_ieee802154_t *)dev,
+                                          opt, value, max_len);
+    }
+}
+
+static int set(netdev_t *dev, netopt_t opt, void *value, size_t value_len)
+{
+    assert(dev);
+
+    DEBUG("set: %s\n", netopt2str(opt));
+
+    switch (opt) {
+        case NETOPT_CHANNEL:
+            assert(value_len == sizeof(uint16_t));
+            set_chan(*((uint16_t *)value));
+            return sizeof(uint16_t);
+        case NETOPT_TX_POWER:
+            assert(value_len == sizeof(int16_t));
+            set_txpower(*((int16_t *)value));
+            return sizeof(int16_t);
+
+        default:
+            return netdev_ieee802154_set((netdev_ieee802154_t *)dev,
+                                         opt, value, value_len);
+    }
+
+}
+
+void isr_radio(void)
+{
+    DEBUG("radio: ISR\n");
+
+    NRF_RADIO->EVENTS_END = 0;
+
+    /* did we just send or receive something? */
+    if (NRF_RADIO->STATE == RADIO_STATE_STATE_RxIdle) {
+        /* only process packet if event callback is set and CRC is valid */
+        if ((nrf802154_dev.netdev.event_callback) &&
+            (NRF_RADIO->CRCSTATUS == 1) &&
+            netdev_ieee802154_dst_filter(&nrf802154_dev, &rxbuf[1])) {
+
+            DEBUG("isr-rx: packet valid -> processing\n");
+
+            txbuf[0] = 5;
+            txbuf[1] = 2;
+            txbuf[2] = 0;
+            txbuf[3] = rxbuf[3];
```

Comment:
this seems dangerous to me. Are we sure, that there is definitely no TX operation in process at this point which could be corrupted by writing in `txbuf`?

Reply:
As fix, we could add a dedicated TX buffer for ACK messages.

Reply:
One more thing: the sending of ACK messages should be conditional, depending if the AUTOACK flag was set. But I would be ok if you add this as a TODO to the implementation status doc.